py/dml/structure.py

```diff
@@ -1987,15 +1989,69 @@ def get_group_registers(group, indexvars):
 
     return regs
 
+def param_linear_int(param):
+    '''If possible, return tuple of coefficients (i0, i1, .., iN, k) such
+    that param's value can be equivalently expressed as
+    (i0*v0 + i1*v1 + ... + iN*vN + k), in DML's 64-bit arithmetic,
+    where vN are index variables. Return None if no such
+    representation can be found.
+
+    If a tuple is returned, then the parameter value either has type
+    uint64 or int64.
+    '''
+    class IndexVar(Expression):
+        type = TInt(64, True)
+        @slotsmeta.auto_init
+        def __init__(self, site, variables): pass
+
+    indices = tuple(IndexVar(param.site,
+                             tuple(1 if x == i else 0
+                                   for x in range(param.dimensions + 1)))
+                    for i in range(param.dimensions))
+    class NotLinear(Exception): pass
+    def expr_linear_int(expr):
+        if isinstance(expr, ctree.Add):
+            return tuple(l + r
+                         for (l, r) in zip(expr_linear_int(expr.lh),
+                                           expr_linear_int(expr.rh)))
+        elif isinstance(expr, ctree.Mult):
+            l = expr_linear_int(expr.lh)
+            r = expr_linear_int(expr.rh)
+            if any(r[:-1]):
+                if any(l[:-1]):
+                    # bilinear
+                    raise NotLinear()
+                # normalize: all variables in 'l'
+                (l, r) = (r, l)
+            return tuple(x * r[-1] for x in l)
+        elif isinstance(expr, ctree.Subtract):
+            return tuple(l - r
+                         for (l, r) in zip(expr_linear_int(expr.lh),
+                                           expr_linear_int(expr.rh)))
+        elif isinstance(expr, ctree.IntegerConstant):
+            return (0,) * param.dimensions + (expr.value,)
+        elif isinstance(expr, IndexVar):
+            return expr.variables
+        elif (isinstance(expr, ctree.Cast) and expr.type.is_int
+              and expr.type.bits == 64):
+            return expr_linear_int(expr.expr)
+        raise NotLinear()
+
+    expr = param.get_expr(indices)
+    try:
+        return tuple(c & ((1 << 64) - 1) for c in expr_linear_int(expr))
+    except (NotLinear, DMLError):
```

Comment:
Isn't this catch on `DMLError` in the wrong place? Shouldn't it be on the `param.get_expr` line?
(Un)fortunately, Python isn't a lazy language :P

Reply:
Right! Which also reveals that I didn't write a test for this corner. Done now.